crates/oxc_linter/src/rules/eslint/no_useless_return.rs

```diff
@@ -0,0 +1,604 @@
+    /// ```js
+    /// function foo() { return; }
+    ///
+    /// function bar() {
+    ///     doSomething();
+    ///     return;
+    /// }
+    ///
+    /// function baz() {
+    ///     if (condition) {
+    ///         doSomething();
+    ///         return;
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// function foo() { return 5; }
+    ///
+    /// function bar() {
+    ///     if (condition) {
+    ///         return;
+    ///     }
+    ///     doSomething();
+    /// }
+    ///
+    /// function baz() {
+    ///     return doSomething();
+    /// }
+    /// ```
+    NoUselessReturn,
+    eslint,
+    pedantic,
+    pending // TODO: implement auto-fix
+);
+
+impl Rule for NoUselessReturn {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        // We only care about return statements without a value
+        let AstKind::ReturnStatement(ret) = node.kind() else {
+            return;
+        };
+
+        // If the return has a value, it's not useless
+        if ret.argument.is_some() {
+            return;
+        }
+
+        // Check if this return statement is useless
+        if Self::is_useless_return(node, ctx) {
+            ctx.diagnostic(no_useless_return_diagnostic(ret.span));
+        }
+    }
+}
+
+impl NoUselessReturn {
+    /// Check if a return statement is useless.
+    fn is_useless_return(return_node: &AstNode, ctx: &LintContext) -> bool {
+        let return_node_id = return_node.id();
+
+        // Check if this return is inside a special control structure
+        for ancestor_id in ctx.nodes().ancestor_ids(return_node_id) {
+            let ancestor = ctx.nodes().get_node(ancestor_id);
+            match ancestor.kind() {
+                // Stop at function boundary
+                AstKind::Function(_) | AstKind::ArrowFunctionExpression(_) => {
+                    break;
+                }
+                // Return inside loops is not useless - it's for early exit
+                AstKind::ForStatement(_)
+                | AstKind::ForInStatement(_)
+                | AstKind::ForOfStatement(_)
+                | AstKind::WhileStatement(_)
+                | AstKind::DoWhileStatement(_) => {
+                    return false;
+                }
+                // Return in finally block is never useless - it overrides other returns
+                AstKind::BlockStatement(block) => {
+                    // Check if this block is a finally block by looking at its parent
+                    let parent_id = ctx.nodes().parent_id(ancestor.id());
+                    let parent = ctx.nodes().get_node(parent_id);
+                    if let AstKind::TryStatement(try_stmt) = parent.kind()
+                        && try_stmt.finalizer.as_ref().is_some_and(|f| f.span == block.span)
+                    {
+                        return false;
+                    }
+                }
```

Comment:
it this always true? or does eslint just not report on this case.

For example:
```ts
function foo(arg0) {
    if (arg0) {
        return "i'm truthy"
    } else {
        return
    }
}